File: starters/python/domains/tasks/service.py
"""tasks domain service.

Depends on core (id encoding + title validation) only. Must never import
domains.users — the boundary enforcer fails the build if it does.

Note: `assign` intentionally does NOT verify that assignee_id refers to a
real user. Doing so would require importing domains.users, which Design D
forbids. Cross-domain checks like that belong in the composition root
(see app/main.py), the only layer allowed to know about both domains.
"""

from core.ids import new_id
from core.validation import validate_title
from domains.tasks.models import Task


class TaskStore:
    """In-memory task store: create/list/assign."""

    def __init__(self):
        self._tasks = {}
        self._counter = 0

    def create(self, title: str) -> Task:
        title = validate_title(title)
        self._counter += 1
        task_id = new_id(self._counter)
        task = Task(id=task_id, title=title)
        self._tasks[task_id] = task
        return task

    def list(self):
        return list(self._tasks.values())

    def get(self, task_id: str) -> Task:
        try:
            return self._tasks[task_id]
        except KeyError:
            raise KeyError(f"no such task: {task_id}") from None

    def assign(self, task_id: str, assignee_id: str) -> Task:
        task = self.get(task_id)
        updated = Task(id=task.id, title=task.title, assignee_id=assignee_id)
        self._tasks[task_id] = updated
        return updated

File: starters/nextjs/src/features/tasks/service.ts
// The `tasks` domain depends on `core` only — never on `users`. Note that
// `assignTask` takes a bare `assigneeId: string`, not a `User`: this domain
// has no way to even express a dependency on `users`.
import { nextId, validateTitle } from "../../core";
import type { Task } from "./types";
import { saveTask, findTask, allTasks } from "./repository";

// @capability tasks.create
export function createTask(input: { title: string }): Task {
  const title = validateTitle(input.title);
  const task: Task = { id: nextId("tsk"), title, status: "open", assigneeId: null };
  return saveTask(task);
}

// @capability tasks.list
export function listTasks(): Task[] {
  return allTasks();
}

// @capability tasks.assign
export function assignTask(taskId: string, assigneeId: string): Task {
  const task = findTask(taskId);
  if (!task) {
    throw new Error(`task not found: ${taskId}`);
  }
  const updated: Task = { ...task, status: "assigned", assigneeId };
  return saveTask(updated);
}
